File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Character> st = new Stack();
        
        for(int i=0;i<num.length();i++){
            char ch = num.charAt(i);
            while(st.size()>0 && st.peek()>ch && k>0){
                st.pop();
                k--;
            }
            st.push(ch);
        }
        
        //if less then k digits are removed
        while(k>0){
            st.pop();
            k--;
        }
        
        //creating answer in reverse
        StringBuilder sb = new StringBuilder();
        while(st.size()!=0){
            sb.append(st.pop());
        }
        
        //removing 0's which will be in front
        while(sb.length()>0 && sb.charAt(sb.length()-1)=='0'){
            sb.deleteCharAt(sb.length()-1);
        }
        
        if(sb.length()==0) return "0";
        
        return sb.reverse().toString();
    }
}

File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        
        int n = num.size();
        stack<int> st;
        
        for(char c : num)
        {
            while(!st.empty() && k > 0 && st.top() > c - '0')
            {
                st.pop();
                k--;
            }
            st.push(c - '0');
        }
        
        while(k > 0 && !st.empty())
        {
            st.pop();
            k--;
        }
        
        string ans;
        while (!st.empty())
        {
            ans += st.top() + '0';
            st.pop();
        }
        
        reverse(ans.begin(), ans.end());
        
        int i = 0;
        while(i < ans.size() && ans[i] == '0')
            i++;
        
        return i == ans.size() ? "0" : ans.substr(i);
    }
};